lmdeploy/pytorch/kernels/pagedattention.py

```diff
@@ -446,22 +604,30 @@ def _kernel_meta():
 
     # shape constraints
     Lq, Lk, Lv = q.shape[-1], k.shape[-1], v.shape[-1]
-    assert Lq == Lk and Lk == Lv
+    assert Lq == Lk, Lv == o.shape[-1]
 
-    sm_scale = 1.0 / (Lq**0.5)
+    if sm_scale is None:
+        sm_scale = 1.0 / (Lq**0.5)
     batch, head = q_seqlens.shape[0], q.shape[-2]
     kv_group_num = q.shape[-2] // k.shape[-2]
 
-    num_warps = 4 if Lk <= 64 else 8
-
-    BLOCK = 64 if k.size(1) < 16 else k.size(1)
+    BLOCK = k.size(1)
+    assert BLOCK >= 16
     BLOCK_DMODEL = triton.next_power_of_2(Lk)
-    num_sub_blocks = BLOCK // k.size(1)
+    if shared_kv:
+        BLOCK_DV = BLOCK_DMODEL
+    else:
+        BLOCK_DV = triton.next_power_of_2(Lv)
+    BLOCK_M = max(16, min(BLOCK, 16384 // BLOCK_DMODEL))
+    if Lk > 512 and BLOCK > 32:
```

Comment:
"and" or "or"?

Reply:
"and".

`Lk>512` => `BLOCK_DMODEL>=1024`
the key smem usage is `BLOCK * BLOCK_DMODEL * sizeof(half)`.